### ISSUE TITLE
Case study of torch.compile / cpp inductor on CPU: min_sum / mul_sum with 1d / matmul-like with static / dynamic shapes

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#106614]
### 🐛 Describe the bug

(I'll add actual benchmarking details and logs and output_code.py in a bit)

I'm doing min_sum and mul_sum in two setups:
1. (D, ) x (D, ) -> scalar
2. (B, N, 1, D) x (B, 1, N, D) -> (B, N, N)

Case (1) is similar to computing some sort of norm of a given input or a distance between two inputs.
Case (2) is matmul-like and is similar to computing distances between all pairs of the batch (mm / pdist / cdist).

When running as `python3 bench_minmul_sum.py --enable-dynamic --avx512 --verbose1`

Findings and questions regarding dynamic shapes:
1. Dynamic / static shape options are not confined to the torch.compile call (which is super-unintuitive and brittle). When using `--enable-dynamic`, all produced `output_code.py` contain dynamic shapes
2. When not using `--enable-dynamic`, all produced `output_code.py` contain static shapes
3. Dynamic shapes `output_code.py` do not record divisibility and always contain an extra tail loop. 
4. Ideally, I would like to have static shapes for 1d ops; static shapes for the inner dim of the 2d ops and dynamic shapes for the batch dims (like can be done for torch.onnx.export). Currently parametrizing static or dynamic shapes is unpredictable :( This is not very good :(
5. Why is `#pragma omp simd simdlen(8)` used? Apparently it's correct, but if it's the register-size-in-float32, I would expect it to be `simdlen(16)`, no? Also, I had hard time to find out the resulting assembly and whether the compiled did loop unrolling (there was no `#pragma omp unroll`). How can one do that? Could the debug option print the produced assembly too? (obtained from `objdump`). Currently there are also too much logs, hard to get through all of them.
6. Better explaining of the tiling story to the users is important - for matmul-like ops especially. And these custom matmul-like ops are quite common in kernel methods and similar (see pykeops)

Findings and questions regarding NaNs:
1. min_sum produced codes have NaN handling
2. It can be good to give an asserting or a hint to the compiler that the inputs will not contain NaNs, so being NaN-compliant is not needed. This NaN handling might be a perf hit for small 1d vectors.

Findings regarding mul_sum (which is equivalent to matrix product or dot product):
1. The mul_sum  pattern is not recognized and `gemm` call or `dot` call is not produced

Findings regarding the benchmarking:
1. On some Linux platforms (e.g. Windows WSLv1) it can be hard to account for CPU throttling. It would be good if PyTorch had some recommendations on how to account for it properly. And on laptops on battery there is often quite severe CPU throttling.

```python
# bench_minmul_sum.py

import os
import sys
import time

enable_dynamic = '--enable-dynamic' in sys.argv

if '--avx512' in sys.argv:
    os.environ['ATEN_CPU_CAPABILITY'] = 'avx512'
if '--verbose1' in sys.argv:
    os.environ['TORCH_COMPILE_DEBUG'] = '1'
if '--verbose2' in sys.argv:
    os.environ['TORCH_LOGS'] = 'output_code'
if '--simd512':
    from torch._inductor import config
    config.cpp.simdlen = 512

from torch._inductor import config
print('config.cpp.simdlen', config.cpp.simdlen)

import torch

min_sum = lambda a, b: torch.min(a, b).sum(-1)
mul_sum = lambda a, b: torch.mul(a, b).sum(-1)

min_sum_compiled_dynamic = torch.compile(min_sum, dynamic = enable_dynamic)
mul_sum_compiled_dynamic = torch.compile(mul_sum, dynamic = enable_dynamic)
min_sum_compiled = torch.compile(min_sum)
mul_sum_compiled = torch.compile(mul_sum)

# 192 is 12 float32x16 (512-bit registers) or 24 float32x8 (256-bit registers)
static_shape = (192,)
dynamic_shape = (6, 183, 192)

def benchmark(name, f, a, b, K = 100):
    tic = time.time()
    for k in range(K):
        f(a, b)
    print(name, a.shape, b.shape, (time.time() - tic) * 1000, 'ms')

# warmup
for k in range(5):
    A = torch.rand(*dynamic_shape, dtype = torch.float32).unsqueeze(2)
    B = torch.rand(*dynamic_shape, dtype = torch.float32).unsqueeze(1)
    a = torch.rand(*static_shape, dtype = torch.float32)
    b = torch.rand(*static_shape, dtype = torch.float32)

    if enable_dynamic:
        min_sum_compiled_dynamic(A, B)
        mul_sum_compiled_dynamic(A, B)
        mul_sum_compiled_dynamic(A, B)
        min_sum_compiled_dynamic(A, A)
    min_sum_compiled(A, B)
    mul_sum_compiled(A, B)
    min_sum_compiled(A, A)
    mul_sum_compiled(A, A)
    min_sum(A, B)
    mul_sum(A, B)
    min_sum(A, A)
    mul_sum(A, A)

    if enable_dynamic:
        min_sum_compiled_dynamic(a, b)
        mul_sum_compiled_dynamic(a, b)
        min_sum_compiled_dynamic(a, a)
        mul_sum_compiled_dynamic(a, a)
    min_sum_compiled(a, b)
    mul_sum_compiled(a, b)
    min_sum_compiled(a, a)
    mul_sum_compiled(a, a)
    min_sum(a, b)
    mul_sum(a, b)
    min_sum(a, a)
    mul_sum(a, a)

# benchmark
A = torch.rand(*dynamic_shape, dtype = torch.float32).unsqueeze(2)
B = torch.rand(*dynamic_shape, dtype = torch.float32).unsqueeze(1)
a = torch.rand(*static_shape, dtype = torch.float32)
b = torch.rand(*static_shape, dtype = torch.float32)

if enable_dynamic:
    benchmark('min_sum_compiled_dynamic_AB', min_sum_compiled_dynamic, A, B)
    benchmark('mul_sum_compiled_dynamic_ab', mul_sum_compiled_dynamic, a, b)
    benchmark('min_sum_compiled_dynamic_AA', min_sum_compiled_dynamic, A, A)
    benchmark('mul_sum_compiled_dynamic_aa', mul_sum_compiled_dynamic, a, a)
benchmark('min_sum_compiled_AB', min_sum_compiled, A, B)
benchmark('mul_sum_compiled_ab', mul_sum_compiled, a, b)
benchmark('min_sum_compiled_AA', min_sum_compiled, A, A)
benchmark('mul_sum_compiled_aa', mul_sum_compiled, a, a)
benchmark('min_sum_AB', min_sum, A, B)
benchmark('mul_sum_ab', mul_sum, a, b)
benchmark('min_sum_AA', min_sum, A, A)
benchmark('mul_sum_aa', mul_sum, a, a)
```

### Versions

2.1.0.dev20230802+cpu

**Comment by jgong5** [comment]
For Inductor c++ codegen specific issues, I noticed that you raised issues related to both perf and UX. How about we focus on perf issues first? I look forward to your performance benchmarking and see what are missing.

> 5\. Why is `#pragma omp simd simdlen(8)` used? Apparently it's correct, but if it's the register-size-in-float32, I would expect it to be `simdlen(16)`, no?

This is only used for the tail handling. We will do explicit vectorization for the main part which doesn't rely on omp simd and for AVX512, it is 16. That's why you see `simdlen(8)` that handles the tail part.

> Better explaining of the tiling story to the users is important - for matmul-like ops especially.

We are not generating code directly for matmul right now. May I know what "tiling story" you are referring to?

> config.cpp.simdlen is 512 despite that without ATEN_CPU_CAPABILITY=avx512, config.show() shows CPU capability usage: AVX2 (which 256-bits registers only). And it seems that this ATEN_CPU_CAPABILITY of only 256-bits is discovered automatically, despite that the laptop supports avx512

We will do dry-compilation to see if we can do avx512. If PyTorch is only built with avx256, only avx256 is enabled here.

I will take PTO this week. I'm copying @EikanWang and @XiaobingSuper for their info.

**Comment by vadimkantorov** [comment]
> If PyTorch is only built with avx256, only avx256 is enabled here.

That's the funny part. It seems that it was built with AVX512 (I'm using stock version), but `os.environ['ATEN_CPU_CAPABILITY']` somehow has effect on reporting the build info (?). So I'm not even sure what `CPU capability` in `torch.__config__.show()` is referring to. And I don't know how to find out whether it's using AVX512 for eager mode.

> How about we focus on perf issues first? I look forward to your performance benchmarking and see what are missing.

I discovered that it's almost impossible to test perf on Windows/WSLv1/Ubuntu as I couldn't find way to tell my laptop's CPU to not throttle, and this platform has no ways of reporting actual CPU freq/perf counters. I should try on Windows/WSLv2 which has access to CPU perf counters to maybe report cycle counts. Do you have any suggestions to set up a benchmark properly?

Regarding unrolling, I hope unrolling is done by the compiler, as for dims 96/192, full unrolling is not so bad. Then is explicit `pragma unroll` helpful? (currently missing)

Regarding tiling, I think when the reduction is done with `sum` (or maybe even with min/max?), one can do gemm-like tiling for the case (2) which is matmul-like, right?

It would be great to have some perf comparison against gemm (for mul-sum case) and against gemm and pykeops (see their example in readme) for a variety of inner dims (pykeops says it's best for <100). Ideally, I would say, PyTorch should be approaching pykeops by speed (including CPU).

<hr>

Yes, the most of other issues/questions I proposed to discuss are indeed UX-related.

Concerning Cpp Inductor, I would suggest that it dumps the actual compiled Assembly next to the C++ code.

And in general IMO we need a way to provide as-detailed-as-we-want hints (including Cpp Inductor config) to torch.compile at per-torch.compile-call basis, not only at the global level.

**Comment by ezyang** [comment]
For the UX question, I have kind of wondered if for operators it makes sense to have some sort of more export-y API, because:

1. You always want fullgraph=True
2. You may want to tweak compile settings in a more fine grained way, without affecting overall torch.compile settings
3. You want careful control over which sizes you end up compiling the function on
4. The majority of guards we install for regular torch.compile aren't relevant; more amenable to a "compile a function, and you get whatever it did at the time you compiled it" semantics

It's a little tricky though if you still want JIT "compile for this shape only when you see it". You might be more accepting of AOT compilation if the results can be cached to disk (see also https://github.com/pytorch/pytorch/pull/101027)

**Comment by vadimkantorov** [comment]
I'm not very knowledgeable about difference of regular compile / export (or what fullgraph=True does), but yes I think it's super-important for the default compile call to be local and allow all sorts of hints passed in (when one wishes).

Compilation artifact cache is super-important. And also super-important to have hooks / ways to completely disable recompilation (and fail if it's needed). Also being able to somehow dump all discovered options / hparams. So that one can easily deploy the same kernels to a new machine and not be bitten by unexpected differently selected impl.

Also very important to somehow minimize amounts of logs and work on their understandability by default.

**Comment by vadimkantorov** [comment]
@ezyang What do you think about adding these norm/matmul-like op variants as some sort of standard benchmark against gemm/pykeops

I don't know how to do this properly, but for limited formulas (maybe pykeops-style), inductor should be able to detect if the reduction is `tilable` or not

min_sum and even mul_argmin are both tilable, aren't they?

**Comment by vadimkantorov** [comment]
E.g. for sub_square_sum would be https://github.com/pytorch/pytorch/issues/28119#issuecomment-624232930, so this dist^p can be implemented with torch.compile

**Comment by vadimkantorov** [comment]
I was meaning, that it's strange that torch.compile detects/uses capability as AVX512 while build info (? or runtime info?) of torch.__config__.show() only displays AVX2

**Comment by sanchitintel** [comment]
Currently, ATen AVX512 kernels are being compiled while building PyTorch, but are not being used by default (the default ATen CPU capability is `avx2`).

As @jgong5 mentioned, after #104165 would land, `torch.__config__.show()` would display CPU capability as `avx512` (on machines that support it), although in eager mode, few ATen kernels would have AVX512 dispatch, and the rest would be dispatched to their AVX2 counterparts, because most ATen CPU kernels (that don't use oneDNN & oneMKL) are slower with AVX512.

**Comment by jgong5** [comment]
> I discovered that it's almost impossible to test perf on Windows/WSLv1/Ubuntu as I couldn't find way to tell my laptop's CPU to not throttle, and this platform has no ways of reporting actual CPU freq/perf counters. I should try on Windows/WSLv2 which has access to CPU perf counters to maybe report cycle counts. Do you have any suggestions to set up a benchmark properly?

Have you tried `time.process_time`?

> Regarding unrolling, I hope unrolling is done by the compiler, as for dims 96/192, full unrolling is not so bad. Then is explicit pragma unroll helpful? (currently missing)

Yes, I believe compiler should be able to do a good job here without explicit pragma unroll.

> Regarding tiling, I think when the reduction is done with sum (or maybe even with min/max?), one can do gemm-like tiling for the case (2) which is matmul-like, right?

Yes, it is essentially a bmm. Why not calling bmm directly?

**Comment by vadimkantorov** [comment]
> Have you tried `time.process_time`?

Will try now, did not know it existed :) Thank you for the hint!

> Yes, it is essentially a bmm. Why not calling bmm directly?

I think mul_sum is good as a test bed for measuring if torch.compile is able to generate tiled loops for matmul-like ops. It probably is an upper bound on the perf difference as bmm is extremely well optimized. So if torch.compile can recognize that the op is matmul-like and starts generating tiled accumulation code, it would be super cool and enable many pekeops/kernel learning-like scenarios.

**Comment by vadimkantorov** [comment]
> Current inductor codegen doesn't handle matmul-like ops yet.

Both min_sum / mul_sum in the example above in the "dynamic shape" part are used in the matmul-like usecase. So it does handle it, but not very efficiently memory-access wise, as it seems tiling is not the best one, right? :(

**Comment by vadimkantorov** [comment]
For these matmul-like tiled codegen, maybe even https://github.com/flame/blis can be used :)

**Comment by vadimkantorov** [comment]
@ezyang Would it be possible to somehow adjust Dynamo so that it accepts torch._dynamo.config optionset as arg (and a way for it to set to disregard the global options; and that options passed to one torch.compile call do not affect the global options)?

Then you could start adding to the core library some convenience functions like https://github.com/pytorch/pytorch/issues/90559 implemented by torch.compile? Or are there other reasons why you don't add such convenience functions decorated as torch.compile into the core?

**Comment by ezyang** [comment]
It should be possible, there's just a lot of wax that needs to be done on this particular topic and we don't have anyone pushing on it full time right now.
